x-pack/plugins/threat_intelligence/public/modules/timeline/components/add_to_timeline/add_to_timeline.tsx

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { memo, useMemo, VFC } from 'react';
+import { DataProvider, QueryOperator } from '@kbn/timelines-plugin/common';
+import { AddToTimelineButtonProps } from '@kbn/timelines-plugin/public';
+import { EuiButtonEmpty, EuiButtonIcon } from '@elastic/eui/src/components/button';
+import { useKibana } from '../../../../hooks/use_kibana';
+import { unwrapValue } from '../../../indicators/lib/unwrap_value';
+import { Indicator, RawIndicatorFieldId } from '../../../../../common/types/indicator';
+import { EMPTY_VALUE } from '../../../../../common/constants';
+
+export interface AddToTimelineProps {
+  /**
+   * Value passed to the timeline. Used in combination with field if is type of {@link Indicator}.
+   */
+  data: Indicator | string;
+  /**
+   * Value passed to the timeline.
+   */
+  field: string;
+  /**
+   * Only used with `EuiDataGrid` (see {@link AddToTimelineButtonProps}).
+   */
+  component?: typeof EuiButtonEmpty | typeof EuiButtonIcon;
+  /**
+   * Used as `data-test-subj` value for e2e tests.
+   */
+  testId?: string;
+}
+
+/**
+ * Add to timeline button, used in many places throughout the TI plugin.
+ * Support being passed a {@link Indicator} or a string, can be used in a `EuiDataGrid` or as a normal button.
+ * Leverages the built-in functionality retrieves from the timeLineService (see ThreatIntelligenceSecuritySolutionContext in x-pack/plugins/threat_intelligence/public/types.ts)
+ * Clicking on the button will add a key-value pair to an Untitled timeline.
+ *
+ * @returns add to timeline button or an empty component.
+ */
+export const AddToTimeline: VFC<AddToTimelineProps> = memo(({ data, field, component, testId }) => {
+  const addToTimelineButton =
+    useKibana().services.timelines.getHoverActions().getAddToTimelineButton;
+
+  const value: string =
+    typeof data === 'string'
+      ? data
+      : unwrapValue(data, field as RawIndicatorFieldId) || EMPTY_VALUE;
+
+  const operator = ':' as QueryOperator;
+  const dataProvider: DataProvider[] = useMemo(
+    () => [
+      {
+        and: [],
+        enabled: true,
+        id: `timeline-indicator-${field}-${value}`,
+        name: value,
+        excluded: false,
+        kqlQuery: '',
+        queryMatch: {
+          field,
+          value,
+          operator,
+        },
+      },
+    ],
+    [field, value]
+  );
+
+  const addToTimelineProps = useMemo(() => {
+    const result: AddToTimelineButtonProps = {
+      dataProvider,
+      field,
+      ownFocus: false,
+    };
+
+    if (component) result.Component = component;
+
+    return result;
+  }, [component, dataProvider, field]);
+
+  return value === EMPTY_VALUE ? (
```

Comment:
does it make sense to return early from the component? 
smth like 
```
if (!value) { return <></> }
```
right after getting the value

Reply:
unless I'm mistaken you shouldn't use useMemo (or any other react hooks) conditionally. By that I mean I believe it's not good to have a early return on a component and have a react hook after that. If that is true I don't think you can return this `<></>` empty component any earlier than it is currently...

Am I mistaken?

Reply:
yes you are right, couple of thoughts:
1. one of the reasons I wasn't really sold on the idea of using memorization extensively without clearly measured performance gains was that it comes with a cost, in this case we need to adjust the way we code to the fact that we use hooks. not really specific to `useMemo` but to all hooks ofc
2. When looking at the component I'm a bit confused with the fact that we wrap the whole component in `memo` HOC and also use `useMemo` inside. If the component is wrapped in `memo` are there gains of using `useMemo` inside it additionally?
3. initially the usage of `EMPTY_VALUE` caught my eye. Do we really need this logic of assigning EMPTY_VALUE and then checking against it? We used EMPTY_VALUE as an UI component for an empty cell, I wouldn't introduce it if it doesn't serve the purpose of displaying `-` somewhere

Reply:
I took the opportunity to read more about `useMemo` vs `React.memo` when to use them when to now.
I feel like this `addToTimeline` component for sure doesn't need to be wrapped by a `memo`.
Also, I feel like that the use of useMemo inside it is also not providing any positive value (I was doing useMemo on the creation of a basic object, not any complex heavy logic..)

I have removed both of them (to be honest the code was lifted off of a Security Solution component). This allows me like you mentioned to [early return](https://github.com/elastic/kibana/pull/138836/files#diff-e6259b848afd323a99380c72301a43f0c4367eca2b5929655f382a78dd793d40R65-R67) if `value` is undefined.

I will be more careful in the future about when to use vs not to use this. I also ned to learn about performance testing with React!
Thanks for challenging me on this!